firestore/lib/src/data/restaurant_provider.dart

```diff
@@ -0,0 +1,168 @@
+
+abstract class RestaurantProvider {
+  Stream<List<Restaurant>> get allRestaurants;
+  Future<void> addRestaurant(Restaurant restaurant);
+  Future<void> addReview(
+      {required String restaurantId, required Review review});
+  void addRestaurantsBatch(List<Restaurant> restaurants);
+  void loadAllRestaurants();
+  void loadFilteredRestaurants(Filter filter);
+  Future<Restaurant> getRestaurantById(String restaurantId);
+  void dispose();
+}
+
+class FirestoreRestaurantProvider implements RestaurantProvider {
+  FirestoreRestaurantProvider() {
+    allRestaurants = _allRestaurantsController.stream;
+  }
+
+  final StreamController<List<Restaurant>> _allRestaurantsController =
+      StreamController();
+
+  @override
+  late final Stream<List<Restaurant>> allRestaurants;
+
+  @override
+  Future<void> addRestaurant(Restaurant restaurant) {
+    final restaurants = FirebaseFirestore.instance.collection('restaurants');
+    return restaurants.add({
+      'avgRating': restaurant.avgRating,
+      'category': restaurant.category,
+      'city': restaurant.city,
+      'name': restaurant.name,
+      'numRatings': restaurant.numRatings,
+      'photo': restaurant.photo,
+      'price': restaurant.price,
+    });
+  }
+
+  @override
+  void addRestaurantsBatch(List<Restaurant> restaurants) {
+    for (var restaurant in restaurants) addRestaurant(restaurant);
+  }
+
+  @override
+  Future<void> addReview(
+      {required String restaurantId, required Review review}) async {
+    final restaurant = await FirebaseFirestore.instance
+        .collection('restaurants')
+        .doc(restaurantId);
+    final newReview = await restaurant.collection('ratings').doc();
+
+    return FirebaseFirestore.instance
+        .runTransaction((Transaction transaction) async {
+      final DocumentSnapshot _restaurantSnapshot =
+          await transaction.get(restaurant);
+      final _restaurant = Restaurant.fromSnapshot(_restaurantSnapshot);
+      final newRatings = _restaurant.numRatings + 1;
+      final newAverage =
+          ((_restaurant.numRatings * _restaurant.avgRating) + review.rating) /
+              newRatings;
+
+      transaction.update(restaurant, {
+        'numRatings': newRatings,
+        'avgRating': newAverage,
+      });
+
+      transaction.set(newReview, {
+        'rating': review.rating,
+        'text': review.text,
+        'userName': review.userName,
+        'timestamp': review.timestamp,
+        'userId': review.userId,
+      });
+    });
+  }
+
+  @override
+  void loadAllRestaurants() {
+    final _querySnapshot = FirebaseFirestore.instance
+        .collection('restaurants')
+        .orderBy('avgRating', descending: true)
+        .limit(50)
+        .snapshots();
+
+    _querySnapshot.listen((event) {
+      final _restaurants = event.docs.map((DocumentSnapshot doc) {
+        return Restaurant.fromSnapshot(doc);
+      }).toList();
+
+      _allRestaurantsController.add(_restaurants);
+    });
```

Comment:
Oh wow, this method actually sets up a listener and keeps our controller in sync. This is definitely one I'd recommend we explain! It'll also be helpful for devs understanding the trade-offs behind this method and deciding whether or not they want to copy the pattern.